internal/crypto/cert.go

```diff
@@ -59,3 +67,94 @@ func ParseCertificates(data []byte, contentType string) (certs []*x509.Certifica
+func MergeCertificateChain(pluginConfig map[string]string, originalCerts []*x509.Certificate) ([]*x509.Certificate, error) {
+	// get cert bundle path
+	if pluginConfig == nil {
+		return originalCerts, nil
+	}
+	certBundlePath, ok := pluginConfig[certBundleKey]
+	if !ok {
+		return originalCerts, nil
+	}
+
+	// validate certificate bundle path
+	basename := filepath.Base(certBundlePath)
+	if !isValidFileName(basename) {
+		return nil, fmt.Errorf("the filename of certificate bundle path is not cross-platform compatible. filename: %s", basename)
+	}
+
+	// read certificate bundle
+	certBundleBytes, err := os.ReadFile(certBundlePath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open user provided cert bundle file %q with err: %w", certBundlePath, err)
+	}
+	certBundle, err := parsePEM(certBundleBytes)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse user provided cert bundle file %q with err: %w", certBundlePath, err)
+	}
+	// merge certificate chain
+	return mergeCertChain(originalCerts, certBundle)
+}
+
+// mergeCertChain appends the parent certificate from the certificate bundle to the original certificate chain.
+// It returns the complete certificate chain along with a nil error if successful, otherwise it returns a nil slice and a non-nil error indicating the failure reason.
+func mergeCertChain(originalCerts, certBundle []*x509.Certificate) ([]*x509.Certificate, error) {
+	if len(originalCerts) == 0 {
+		return nil, errors.New("the certificate chain accessed from Azure Key Vault is empty")
+	}
+	if len(certBundle) == 0 {
+		return nil, errors.New("the certificate bundle file is empty or in a unsupported format")
+	}
+	// leaf certificate
+	leafCert := originalCerts[0]
+
+	// generate certificate pools
+	rootPool := x509.NewCertPool()
+	intermediatePool := x509.NewCertPool()
+
+	// verify options
+	opts := x509.VerifyOptions{
+		Roots:         rootPool,
+		Intermediates: intermediatePool,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+	}
+
+	// try to validate the certificate chain by using the original certificates
+	// accessed from AKV.
+	for _, cert := range originalCerts[1:] {
+		if bytes.Equal(cert.RawIssuer, cert.RawSubject) {
+			rootPool.AddCert(cert)
+		} else {
+			intermediatePool.AddCert(cert)
+		}
+	}
+	// verify and build certificate chain for leaf CA.
+	// skip the error because it will try to use cert bundle to fix it.
+	if certChains, err := leafCert.Verify(opts); err == nil {
+		return certChains[0], nil
+	}
+
+	// if the original certificates are not enough for generating the chain,
+	// try to use the certificate bundle.
+	for _, cert := range certBundle {
+		if bytes.Equal(cert.RawIssuer, cert.RawSubject) {
+			rootPool.AddCert(cert)
+		} else {
+			intermediatePool.AddCert(cert)
+		}
+	}
+	certChains, err := leafCert.Verify(opts)
+	if err != nil {
+		return nil, fmt.Errorf("cannot merge the certificate chain by the certificate bundle. error: %w", err)
+	}
+	return certChains[0], nil
```

Comment:
Looks like this logic contains one piece of small logic (leafCert.Verify) and repeated twice. Can we have a function for it given we might also call this function at top of `MergeCertificateChain` when `pluginConfig` is missing.

Reply:
Refactored.